Address field: Add super-scaffolding support, document new field and document updating dependent-fields in a form

pascallaliberte opened the pull request:
Closes #533

This PR adds super-scaffolding support, documents the field and documents the self-updating form mechanism to be used elsewhere in forms.

* [x] Add the new `address_field` to the super-scaffolding mechanism
* [x] Document the new `address_field` compound field partial, its large support for country and region choices out of the box.
* [x] Add the address field to the showcase preview
* [x] Document the `attributes/address` partial and its display options (one_line)
* [x] Clarify the options available for the address_field (can it be read-only? will that render all sub-fields read-only?). See the showcase preview
* [x] Document the new mechanism for auto-updating dependent form fields (e.g. the states/provinces/region field updates based on `country_id` selected). Resolves https://github.com/bullet-train-co/bullet_train-fields/issues/42

pascallaliberte commented:
I'm trying to find a good name, a good file name / title, for the new pattern introduced with the `address_field` whereby the value of one field updates parts of the form further down (a section or a single field).

The trick uses:
* Two Stimulus controllers (`dependable_controller.js` and `refresh_fields_controller.js`),
* a turbo_frame around the fields expected to be updated via a URL refetch, and
* a helper (`accept_query_string_override_for(form, method)`) to avoid needing to manually clear the field name through a `strong_params` at the controller-level.

Some names I've got down:
1. Turbo Reactive Field Changes `turbo-reactive-field-changes`
2. Self-Updating Form Fields `self-updating-form-fields`
3. Turbo Conditional Field Updates `turbo-conditional-field-updates`
4. Magic Dependent Fields `magic-dependent-fields`

Ideas? @kaspth @jagthedrummer @julianrubisch

gazayas commented:
@pascallaliberte I personally like `turbo-reactive-field-changes`!

Looking at the code here...

https://github.com/bullet-train-co/bullet_train-core/blob/ffe2aa34e9dd4a881711ffb67ebda41a243c98da/bullet_train-themes-tailwind_css/app/views/themes/tailwind_css/fields/_address_field.html.erb#L41-L48

I think the name `turbo-reactive-field-changes` makes it easy to understand that the target of the change is a turbo frame, and also since we're doing changes with Stimulus controllers, the use of the word `reactive` here sits well with me too.

## Other thoughts
If I would change anything, I might call it `turbo-reactive-region-field-changes` to be more specific. I know it's two fields, `:region_id` and `:postal_code`, but I feel like `:postal_code` is closely tied to region so we can pass for covering this with simply `region` in the title (for the sake of being specific concerning what we're changing).


## End goal 
> I'm trying to find a good name, a good file name / title, for the new pattern introduced with the address_field

I'm not sure I understand what the end goal is, do we want a new name for a file that we're adding in `bullet_train-core`, or a name for the purposes of adding it to the documentation? (Looking at the title of the PR I'm assuming it's mainly for documentation purposes.) Sorry if this was clear in your comment and I missed it. Either way, I like `turbo-reactive-field-changes` :smiley:

pascallaliberte commented:
@gazayas Thanks for chiming in! I appreciate it.

The end goal: document the pattern for uses outside the address field country > region + postal code field updates.

So I'm looking for naming the pattern, creating a page in the docs with a title and a short-ish slug. Something catchy. This pattern is a Bullet Train first, being a super-scaffold friendly no-config way to update fields on a form based on another field. I also think it's a Rails first, so I think it's worth giving BT a marketing boost from it, and hence giving it a solid name.

I also like "Turbo Reactive Field Changes". One drawback: I think it's missing the magic that `accept_query_string_override_for` provides, where you don't need to modify the strong_params in the controller to make sure the form can read from a query string param.

It's not just that Turbo is involved. It's that this technique is low-config.

Thoughts?

jagthedrummer commented:
Of the proposed names I like "Magic Dependent Fields" the best. Partly because it's the shortest, and partly because I think it best describes _what_ the feature is about, without the name being pegged to _how_ it does it. I'd maybe also drop "Magic" and just call it "Dependent Fields".

kaspth commented:
I'm having a bit of a hard time understanding the underlying API here, it seems like we're missing some wiring somewhere particularly given `accept_query_string_override_for`. I wish we had a form controller to abstract the dependencies away. Something like this maybe?

```erb
<%= form_with model: @tangible_thing, data: { controller: "form" } do |form| %>
  <%# Ensure we fire off a request on change, but route it into the dedicated frame %>
  <%= form.select :country_id, data: { action: "change->form#requestSubmit", turbo_frame: form.refreshed_field_name(:country_id) } %>

  <%= turbo_frame_tag form.refreshed_field_name(:country_id) do %>
    <%= render "addresses/form" %>
  <% end %>
<% end %>
```

pascallaliberte commented:
@kaspth I appreciate this. I'll think it over some more.

pascallaliberte commented:
@kaspth In our case, I was looking for a way that would fit a few constraints:
1. not having to add an attribute to the form wrapping the fields because of super-scaffolding support (in particular in the context of adding the address field to an existing form)
2. not requiring any changes to the resource controller, such as reading from the query_string params and setting a form object's value
3. not requiring any validation skipping from side-effects of submitting the form

But the API could be cleaned up, to your credit. I did the following changes:

* Abstracted `accept_query_string_override_for` into a `_dependent_fields_frame` partial
* Added documentation and clarified down to two concepts (see docs): 1) Dependent Fields Pattern and 2) Dependent Fields Frame
* Chose the main heading of "Dynamic Forms and Dependent Fields"
* Changed the Stimulus controllers to match the concepts: `dependable_controller.js` (unchanged) and `dependent_fields_frame_controller.js` (from refresh_fields_controller.js).
* Added a showcase preview for `_dependent_fields_frame`.

@gazayas, @jagthedrummer: I'm dropping "Turbo" and "Magic" from the name of this feature because of how the two concepts (Dependent Fields Pattern and Dependent Fields Frame) are explained, it really feels more like a sober toolkit than a central innovation. It's also found within a heading of "Dynamic Forms and Dependent Fields" in the docs, for scalability and "obviousness"(?).

Reactions? Rebuttals?

gazayas commented:
@pascallaliberte Sounds good to me! After reading @jagthedrummer's comment too a naming convention that has a focus on `dependent` makes sense for the logic in place, so I'm currently ok with "Dependent Fields Pattern" and "Dependent Fields Frame"

gazayas commented on bullet_train-fields/app/javascript/controllers/dependent_fields_frame_controller.js:
```diff
@@ -8,7 +8,7 @@ export default class extends Controller {
   static classes = [ "loading" ]
   
-  updateFrameFromDependentField(event) {
+  updateFrameFromDependableField(event) {
```
Am I right in assuming that `DependableField` represents the main field being triggered which then updates the `DependentField`? Just want to make sure I'm getting these naming conventions right.

pascallaliberte replied:
You got it 👍

kaspth commented:
@pascallaliberte great job! I just skimmed the code and docs and it looks a lot clearer to me.

kaspth commented on bullet_train-themes-light/app/views/showcase/previews/field_partials/attribute_partials/_about_attribute_partials.html.erb:
```diff
@@ -0,0 +1,19 @@
+  Read more about them and about the form field partials in the extensive <a href="https://bullettrain.co/docs/field-partials" target="_blank">developer documentation</a>.
+<% end %>
+
+<% if partial.body? %>
+  <%= partial.body.yield Scaffolding::CompletelyConcrete::TangibleThing.new %>
+<% end %>
```
We've also got an `optional` proxy that only calls the method if `body?`, like this:

```suggestion
<%= partial.body.optional.yield Scaffolding::CompletelyConcrete::TangibleThing.new %>
```

kaspth commented on bullet_train-themes-light/app/views/showcase/previews/field_partials/attribute_partials/_about_attribute_partials.html.erb:
```diff
@@ -0,0 +1,19 @@
+<% showcase.options do |o| %>
+  <% o.required :object, "Reference to the model object", type: "Object" %>
+  <% o.required :attribute, "Attribute of the model" %>
+  <% if partial.options? %>
+    <%= partial.options.yield o %>
+  <% end %>
```
```suggestion
  <%= partial.options.optional.yield o %>
```

kaspth commented on bullet_train-themes-light/app/views/showcase/previews/field_partials/attribute_partials/_address.html.erb:
❤️

kaspth commented on bullet_train/docs/field-partials/dynamic-forms-dependent-fields.md:
```diff
@@ -0,0 +1,123 @@
+<%= render "shared/fields/dependent_fields_frame", 
+  id: form.field_id(:heard_from, :dependent_fields),
+  form: form,
+  dependable_fields: [:heard_from] %>
```
Are we missing a `do` here?

```suggestion
  dependable_fields: [:heard_from] do %>
```

kaspth commented on bullet_train/docs/field-partials/dynamic-forms-dependent-fields.md:
````diff
@@ -0,0 +1,123 @@
+We'll move the conditional on the `disabled` property. And we'll also let the `dependent_fields_frame` underlying controller handle disabling the field automatically when the `turbo_frame` awaits updates.
+
+```erb
+<%= render "shared/fields/dependent_fields_frame", 
````
Technically not relevant to this PR, but I keep thinking I should add a default Form Builder for Bullet Train, so we can get stuff like this out of the box:

```erb
<%= form.dependent_fields_frame :heard_from do |controller| %>
  <%= form.text_field :heard_from_other, disabled: form.object&.heard_from != "other",
      data: { "#{controller}-target": "field" %>
<% end %>
```

kaspth commented on bullet_train/docs/field-partials/dynamic-forms-dependent-fields.md:
I think this documentation is really clear!

jagthedrummer reviewed (approved):
I really like how this turned out. Great write up of everything. Nice work, @pascallaliberte!
